src/index.tsx

```diff
@@ -0,0 +1,185 @@
+  onError?: (error: Error, info: {componentStack: string}) => void
+  resetKeys?: Array<unknown>
+  FallbackComponent: React.ComponentType<FallbackProps>
+}
+
+declare function FallbackRender(
+  props: FallbackProps,
+): React.ReactElement<
+  unknown,
+  string | React.FunctionComponent | typeof React.Component
+> | null
+
+interface ErrorBoundaryPropsWithRender {
+  onResetKeysChange?: (
+    prevResetKeys: Array<unknown>,
+    resetKeys: Array<unknown>,
+  ) => void
+  onReset?: (...args: Array<unknown>) => void
+  onError?: (error: Error, info: {componentStack: string}) => void
+  resetKeys?: Array<unknown>
+  fallbackRender: typeof FallbackRender
+}
+
+interface ErrorBoundaryPropsWithFallback {
+  onResetKeysChange?: (
+    prevResetKeys: Array<unknown>,
+    resetKeys: Array<unknown>,
+  ) => void
+  onReset?: (...args: Array<unknown>) => void
+  onError?: (error: Error, info: {componentStack: string}) => void
+  resetKeys?: Array<unknown>
+  fallback: React.ReactElement<
+    unknown,
+    string | React.FunctionComponent | typeof React.Component
+  > | null
+}
+
+type ErrorBoundaryProps =
+  | ErrorBoundaryPropsWithFallback
+  | ErrorBoundaryPropsWithComponent
+  | ErrorBoundaryPropsWithRender
+
+type ErrorBoundaryState = {error: Error | null}
+const initialState = {error: null}
+class ErrorBoundary extends React.Component<
+  // TODO: why is there not a PropsWithRefAndChildren? 🤔
+  React.PropsWithRef<ErrorBoundaryProps>,
+  ErrorBoundaryState
+> {
+  static getDerivedStateFromError(error: Error) {
+    return {error}
+  }
+
+  state = initialState
+  updatedWithError = false
+  resetErrorBoundary = (...args: Array<unknown>) => {
+    this.props.onReset?.(...args)
+    this.reset()
+  }
+
+  reset() {
+    this.updatedWithError = false
+    this.setState(initialState)
+  }
+
+  componentDidCatch(error: Error, info: React.ErrorInfo) {
+    this.props.onError?.(error, info)
+  }
+
+  componentDidUpdate(prevProps: ErrorBoundaryProps) {
+    const {error} = this.state
+    const {resetKeys} = this.props
+
+    // There's an edge case where if the thing that triggered the error
+    // happens to *also* be in the resetKeys array, we'd end up resetting
+    // the error boundary immediately. This would likely trigger a second
+    // error to be thrown.
+    // So we make sure that we don't check the resetKeys on the first call
+    // of cDU after the error is set
+
+    // TODO: figure out why this is considered an unnecessary condition...
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (error !== null && !this.updatedWithError) {
+      this.updatedWithError = true
+      return
+    }
+
+    // TODO: figure out why this is considered an unnecessary condition...
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (error !== null && changedArray(prevProps.resetKeys, resetKeys)) {
+      this.props.onResetKeysChange?.(
+        prevProps.resetKeys as Array<unknown>,
```

Comment:
I wonder if making type generic in the type of reset keys (like`Array<T>`) would help with type casting. 

I'm of the opinion those should be used sparingly.

Reply:
Actually, I figured out another way to fix this that didn't require type casting 👍